app/models/listing.rb

```diff
@@ -10,7 +13,81 @@ def self.pending
     where(are_interested: nil)
   end
 
+  def self.for_settings(user_settings)
+    transit_max_time(user_settings.transit_max_time)
+      .walk_score_minimum(user_settings.walk_score_minimum)
+      .transit_score_minimum(user_settings.transit_score_minimum)
+      .bike_score_minimum(user_settings.transit_score_minimum)
+      .personal_crime_grade_minimum(user_settings.personal_crime_grade_minimum)
+      .property_crime_grade_minimum(user_settings.property_crime_grade_minimum)
+  end
+
+  def self.order_by_price_and_transit_time
+    order(price: :asc, transit_travel_in_minutes: :asc)
+  end
+
   def self.bart_max_distance(distance)
     where("bart_distance is null OR bart_distance <= ?", distance)
   end
+
+  def self.transit_max_time(minutes)
+    where(
+      "transit_travel_in_minutes is null OR transit_travel_in_minutes <= ?",
+      minutes
+    )
+  end
+
+  def self.walk_score_minimum(minutes)
+    score_minimum(:walk, minutes)
+  end
+
+  def self.transit_score_minimum(minutes)
+    score_minimum(:transit, minutes)
+  end
+
+  def self.bike_score_minimum(minutes)
+    score_minimum(:bike, minutes)
+  end
+
+  def self.personal_crime_grade_minimum(grade)
+    grade_minimum(:personal, grade)
+  end
+
+  def self.property_crime_grade_minimum(grade)
+    grade_minimum(:property, grade)
+  end
+
+  def user_settings
+    user.settings
+  end
+
+  private
+
+  def self.score_minimum(type, value)
+    where("#{type}_score is null OR #{type}_score >= ?", value)
+  end
+
+  def self.grade_minimum(type, grade)
+    key = "#{type}_crime_grade"
+    accepted_grades = [grade] + higher_grades(grade)
+    query_formatted_grades = accepted_grades.map { |g| "'#{g}'" }.join(', ')
+    # SQL-injection vulnerability!!!
+    # Not sure of a good way to generate this query
+    where("#{key} is null OR #{key} IN (#{query_formatted_grades})")
```

Comment:
Rails 5.1 came out since this PR was made, so:

```ruby
where(key => nil).or(where(key => accepted_grades))
```

And `query_formatted_grades` can go away 😇